### PR TITLE
[aot autograd] avoid cloning some inputs unnecessarily when they dont require grad

### PR DESCRIPTION
**Pull request by bdhirsh**
When constructing the joint graph, we normally have to clone any inputs that are mutated, so that we can pass in the original, pre-mutation inputs as leaves to autograd.

Previously, we were doing this for all mutated inputs - but we only need to do it for inputs that require gradients and participate in autograd.

Hopefully this should speed up code like batch norm - I think before this we were unnecessarily cloning the running stats during training.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #96784
* __->__ #96342
* #96341
* #96340
* #96339
* #96235
* #95992

### CODE REVIEW

**Comment by albanD** [comment]
Why is that ok to do? Inputs that don't require gradients might still be used during backward.

**Comment by bdhirsh** [comment]
@albanD I'll try to clarify (but also tell let me know if this sounds wrong). The situation where we used to clone but no longer do in this PR is the following:

```
def f(a, b):
    a.mul_(2)
    b.mul_(3)
    return a + b

# First arg requires grad, second doesn't
a = torch.ones(2, requires_grad=True)
b = torch.ones(2, requires_grad=False)
```

When we call `autograd.grad()` to trace out a joint graph, we used to do the following:

```
# This is effectively the joint function that we functionalize and trace in AOTAutograd
def joint_function(fn, a, b, out_tangent)
    # Previously, we would notice that both a and b were mutated inputs,
    # so we would .clone() blindly clone all inputs that are mutated (in case they are passed as grad_primals to autograd.grad())
    a_clone = a.clone()
    b_clone = b.clone()
    out = fn(a_clone, b_clone)
    # But in this example, we only needed to clone "a" (so we can pass it as a grad_primal before the mutation),
    # while "b" isn't used as a grad_primal in `autograd.grad()`.
    backward_out = torch.autograd.grad(
        [out],
        [a],
        grad_outputs=[out_tangent],
        allow_unused=True,
    )
    return [out], [backward_out[0], None]
```

The only other situation I can think of where "b" might be used during the backward is if "b" were saved-for-backward. But AOTAutograd executes the entire forward graph before saving any tensors for backward - if it happened to save a graph input (or anything else) that got a mutation, the mutation will have already happened at that point.

**Comment by albanD** [comment]
Thanks for the detailed explanation. If the concern is for the Tensor you end up giving to autograd.grad, then I do agree yes.

File: torch/_functorch/aot_autograd.py [hunk 1]
```diff
@@ -866,14 +866,17 @@ class AOTConfig:
...
-    if meta.input_info[idx].mutates_data:
-        # Make sure the primal we pass to autograd.grad()
-        # sees the tensor before the mutation
-        return t.clone()
-    if meta.input_info[idx].mutates_metadata:
-        # Make sure the primal we pass to autograd.grad()
-        # sees the tensor before the metadata mutation
-        return t.view(t.shape)
+    if idx in meta.mutated_inp_indices:
+        # We only need to bother cloning mutated inputs that participate in autograd.
```
Comment (albanD) [inline torch/_functorch/aot_autograd.py:870]: nit: We only need to clone mutated inputs for which we need the pre-mutation gradient.

**Review by albanD — APPROVED** [review]
SGTM

**Review by ezyang — APPROVED** [review]
